Modules/_testinternalcapi/set.c

```diff
@@ -0,0 +1,83 @@
+#include "parts.h"
+#include "../_testcapi/util.h"  // NULLABLE, RETURN_INT
+
+#include "pycore_setobject.h"
+
+
+static PyObject *
+set_update(PyObject *self, PyObject *args)
+{
+    PyObject *set, *iterable;
+    if (!PyArg_ParseTuple(args, "OO", &set, &iterable)) {
+        return NULL;
+    }
+    NULLABLE(set);
+    NULLABLE(iterable);
+    RETURN_INT(_PySet_Update(set, iterable));
+}
+
+static PyObject *
+set_next_entry(PyObject *self, PyObject *obj)
+{
+    NULLABLE(obj);
+    PyObject *result = PyList_New(0);
+    if (result == NULL) {
+        return NULL;
+    }
+
+    int set_next = 0;
+    Py_ssize_t i = 0, count = 0;
+    Py_hash_t h;
+    PyObject *item;
+    while ((set_next = _PySet_NextEntry(obj, &i, &item, &h))) {
+        if (set_next == -1 && PyErr_Occurred()) {  // obj is not a set
+            goto error;
+        }
+
+        count++;
+        PyObject *index = PyLong_FromSsize_t(count);
+        if (index == NULL) {
+            goto error;
+        }
+        PyObject *hash = PyLong_FromSize_t((size_t)h);
+        if (hash == NULL) {
+            Py_DECREF(index);
+            goto error;
+        }
+        PyObject *tup = PyTuple_Pack(3, index, item, hash);
```

Comment:
You can use `Py_BuildValue("nOn", count, item, (Py_ssize_t)h)`. It will save 11 lines of code.

Reply:
On other hand, why return `count`?

Reply:
I would write test similarly to the test for `PyDict_Next()` -- implement a wrapper for single step and call it repeatedly in Python code.